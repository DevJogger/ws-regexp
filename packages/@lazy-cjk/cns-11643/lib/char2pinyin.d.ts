import { EnumPinyinMode, IZhuyin2PinyinTableRow, IZhuyin2PinyinTable, EnumPinyinType } from './types';
export declare function zhuyin2pinyin_table(pinyinMode?: EnumPinyinMode): IZhuyin2PinyinTable;
export declare function zhuyin2pinyin(zhuyin: string, pinyinMode?: EnumPinyinMode): IZhuyin2PinyinTableRow;
export declare function uni2pinyin(uni: string | number, pinyinMode?: EnumPinyinMode): IZhuyin2PinyinTableRow;
export declare function uni2pinyinTypeValue(uni: string | number, pinyinMode?: EnumPinyinMode, pinyinType?: EnumPinyinType): string;
export declare function char2pinyin(char: string, pinyinMode?: EnumPinyinMode): IZhuyin2PinyinTableRow;
export declare function char2pinyinTypeValue(char: string, pinyinMode?: EnumPinyinMode, pinyinType?: EnumPinyinType): string;
export default char2pinyin;
