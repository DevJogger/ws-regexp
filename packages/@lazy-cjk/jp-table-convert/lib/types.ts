
export const KEY_JP = 'jp' as const;
export const KEY_ZHT = 'zht' as const;
export const KEY_ZHS = 'zhs' as const;

export type IKEY_FROM_TO = typeof KEY_JP | typeof KEY_ZHT | typeof KEY_ZHS | 'jp' | 'zht' | 'zhs';

export interface ITABLE_MAIN
{
	[KEY_JP]: ITABLE,
	[KEY_ZHT]: ITABLE,
	[KEY_ZHS]: ITABLE,
}

export interface ITABLE
{
	[key: string]: ITABLESUB
}

export interface ITABLESUB
{
	[KEY_JP]: string,
	[KEY_ZHT]: string,
	[KEY_ZHS]: string,
}

export interface IOptions
{
	/**
	 * 忽略的字 or 任何支援 indexOf 的 Object
	 */
	skip?,
	/**
	 * safe mode
	 */
	safe?: boolean,
}

export interface IFrom2To extends Function
{
	(str, options?: IOptions): string;
}

