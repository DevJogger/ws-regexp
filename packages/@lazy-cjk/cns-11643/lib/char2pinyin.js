"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.char2pinyinTypeValue = exports.char2pinyin = exports.uni2pinyinTypeValue = exports.uni2pinyin = exports.zhuyin2pinyin = exports.zhuyin2pinyin_table = void 0;
const uni2pinyin_01_1 = __importStar(require("./uni2pinyin_01"));
const uni2pinyin_02_1 = __importStar(require("./uni2pinyin_02"));
const uni2zhuyin_1 = require("./util/uni2zhuyin");
function zhuyin2pinyin_table(pinyinMode) {
    const zhuyin2pinyinMode = pinyinMode ? uni2pinyin_01_1.zhuyin2pinyin_01_table : uni2pinyin_02_1.zhuyin2pinyin_02_table;
    return zhuyin2pinyinMode();
}
exports.zhuyin2pinyin_table = zhuyin2pinyin_table;
function zhuyin2pinyin(zhuyin, pinyinMode) {
    const zhuyin2pinyinMode = pinyinMode ? uni2pinyin_01_1.zhuyin2pinyin_01 : uni2pinyin_02_1.zhuyin2pinyin_02;
    return zhuyin2pinyinMode(zhuyin);
}
exports.zhuyin2pinyin = zhuyin2pinyin;
function uni2pinyin(uni, pinyinMode) {
    const uni2pinyinMode = pinyinMode ? uni2pinyin_01_1.default : uni2pinyin_02_1.default;
    return uni2pinyinMode(uni);
}
exports.uni2pinyin = uni2pinyin;
function uni2pinyinTypeValue(uni, pinyinMode, pinyinType) {
    pinyinType = uni2zhuyin_1.handlePinyinType(pinyinType);
    return uni2pinyin(uni, pinyinMode)[pinyinType];
}
exports.uni2pinyinTypeValue = uni2pinyinTypeValue;
function char2pinyin(char, pinyinMode) {
    const char2pinyinMode = pinyinMode ? uni2pinyin_01_1.char2pinyin_01 : uni2pinyin_02_1.char2pinyin_02;
    return char2pinyinMode(char);
}
exports.char2pinyin = char2pinyin;
function char2pinyinTypeValue(char, pinyinMode, pinyinType) {
    pinyinType = uni2zhuyin_1.handlePinyinType(pinyinType);
    return char2pinyin(char, pinyinMode)[pinyinType];
}
exports.char2pinyinTypeValue = char2pinyinTypeValue;
exports.default = char2pinyin;
//# sourceMappingURL=char2pinyin.js.map