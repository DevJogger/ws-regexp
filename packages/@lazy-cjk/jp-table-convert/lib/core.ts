import { ITABLE_MAIN, ITABLE, KEY_JP, KEY_ZHT, KEY_ZHS, ITABLESUB, IKEY_FROM_TO } from './types';
import { IKanjiComparisonTable } from '@lazy-cjk/jp-table-comparison/lib/types';

export function _build_record(src: IKanjiComparisonTable): ITABLE_MAIN
{
	let to = {} as ITABLE_MAIN;
	to[KEY_JP] = {} as ITABLE;
	to[KEY_ZHT] = {} as ITABLE;
	to[KEY_ZHS] = {} as ITABLE;

	src.forEach(function (vrow)
	{
		let [jp, zht, zhs] = vrow;

		let k: IKEY_FROM_TO = KEY_JP;
		for (let c of jp)
		{
			if (!c || to[k][c])
			{
				continue;
			}

			to[k][c] = to[k][c] || {} as ITABLESUB;

			to[k][c][k] = c;
			to[k][c][KEY_ZHT] = zht[0];
			to[k][c][KEY_ZHS] = zhs[0];
		}

		k = KEY_ZHT;
		for (let c of zht)
		{
			if (!c || to[k][c])
			{
				continue;
			}

			to[k][c] = to[k][c] || {} as ITABLESUB;

			to[k][c][KEY_JP] = jp[0];
			to[k][c][k] = c;
			to[k][c][KEY_ZHS] = zhs[0];
		}

		k = KEY_ZHS;
		for (let c of zhs)
		{
			if (!c || to[k][c])
			{
				continue;
			}

			to[k][c] = to[k][c] || {} as ITABLESUB;

			to[k][c][KEY_JP] = jp[0];
			to[k][c][KEY_ZHT] = zht[0];
			to[k][c][k] = c;
		}
	});

	return to;
}

export function _build_table(ZHJP_TABLE: IKanjiComparisonTable, ZHJP_TABLE_SAFE: IKanjiComparisonTable)
{
	const TABLE = _build_record(ZHJP_TABLE);
	const TABLE_SAFE = _build_record(ZHJP_TABLE_SAFE);

	return {
		TABLE,
		TABLE_SAFE,
	}
}
