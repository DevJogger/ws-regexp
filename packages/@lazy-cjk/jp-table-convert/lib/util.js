"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._getdata = void 0;
const table_1 = require("./table");
function _getdata(char, from, to, safe) {
    if (safe) {
        return (table_1.TABLE_SAFE[from][char]) ? table_1.TABLE_SAFE[from][char][to] : null;
    }
    return (table_1.TABLE[from][char]) ? table_1.TABLE[from][char][to] : null;
}
exports._getdata = _getdata;
//# sourceMappingURL=util.js.map