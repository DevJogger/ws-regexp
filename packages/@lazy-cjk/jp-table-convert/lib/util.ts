import { IKEY_FROM_TO } from './types';
import { TABLE_SAFE, TABLE } from './table';

export function _getdata(char: string, from: IKEY_FROM_TO, to: IKEY_FROM_TO, safe?: boolean): string
{
	return (safe ? TABLE_SAFE : TABLE)[from][char]?.[to];
}
