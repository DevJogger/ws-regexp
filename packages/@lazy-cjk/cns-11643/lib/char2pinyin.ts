import uni2pinyin_01, { char2pinyin_01, zhuyin2pinyin_01, zhuyin2pinyin_01_table } from './uni2pinyin_01';
import uni2pinyin_02, { char2pinyin_02, zhuyin2pinyin_02_table, zhuyin2pinyin_02 } from './uni2pinyin_02';
import { EnumPinyinMode, IZhuyin2PinyinTableRow, IZhuyin2PinyinTable, EnumPinyinType } from './types';
import { handlePinyinType } from './util/uni2zhuyin';

export function zhuyin2pinyin_table(pinyinMode?: EnumPinyinMode)
{
	const zhuyin2pinyinMode = pinyinMode ? zhuyin2pinyin_01_table : zhuyin2pinyin_02_table;

	return zhuyin2pinyinMode()
}

export function zhuyin2pinyin(zhuyin: string, pinyinMode?: EnumPinyinMode)
{
	const zhuyin2pinyinMode = pinyinMode ? zhuyin2pinyin_01 : zhuyin2pinyin_02;

	return zhuyin2pinyinMode(zhuyin)
}

export function uni2pinyin(uni: string | number, pinyinMode?: EnumPinyinMode)
{
	const uni2pinyinMode = pinyinMode ? uni2pinyin_01 : uni2pinyin_02;

	return uni2pinyinMode(uni)
}

export function uni2pinyinTypeValue(uni: string | number, pinyinMode?: EnumPinyinMode, pinyinType?: EnumPinyinType)
{
	pinyinType = handlePinyinType(pinyinType)

	return uni2pinyin(uni, pinyinMode)[pinyinType]
}

export function char2pinyin(char: string, pinyinMode?: EnumPinyinMode)
{
	const char2pinyinMode = pinyinMode ? char2pinyin_01 : char2pinyin_02;

	return char2pinyinMode(char)
}

export function char2pinyinTypeValue(char: string, pinyinMode?: EnumPinyinMode, pinyinType?: EnumPinyinType)
{
	pinyinType = handlePinyinType(pinyinType)

	return char2pinyin(char, pinyinMode)[pinyinType]
}

export default char2pinyin
