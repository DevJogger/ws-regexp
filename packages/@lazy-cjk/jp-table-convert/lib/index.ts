import { IFrom2To, IOptions, KEY_ZHT, KEY_ZHS, KEY_JP } from './types';
import { TABLE, TABLE_KEYS } from './table';
import { _re_cjk_conv } from 'regexp-helper/lib/cjk-conv';
import { UString } from 'uni-string';
import { _getdata } from './util';

// /[\u4E00-\u9FFF\u{20000}-\u{2FA1F}]+/u
const REGEXP_TEST = new RegExp(_re_cjk_conv('u').source + '+', 'u');

const defaultOptions: IOptions = {
	safe: true,
};

namespace _
{
	export let jp2zht: IFrom2To;
	export let jp2zhs: IFrom2To;
	export let zht2jp: IFrom2To;
	export let zht2zhs: IFrom2To;
	export let zhs2jp: IFrom2To;
	export let zhs2zht: IFrom2To;

	let langs = Object.keys(TABLE);

	// @ts-ignore
	langs.forEach(function (from: IKEY_FROM_TO)
	{
		// @ts-ignore
		langs.forEach(function (to: IKEY_FROM_TO)
		{
			if (from == to) return;

			_[`${from}2${to}`] = function (str, options?: IOptions): string
			{
				if (!REGEXP_TEST.test(str.toString()))
				{
					return str;
				}

				options = Object.assign({}, defaultOptions, options);

				return UString.split(str, '')
					.map(function (char: string)
					{
						if (options.skip && options.skip.indexOf(char) != -1)
						{
							return char;
						}

						let c: string;
						if (c = _getdata(char, from, to, options.safe))
						{
							return c;
						}

						return char;
					})
					.join('')
					;
			} as IFrom2To;
		});
	});
}

/**
 * only 只將 日文漢字 => 轉為繁漢字
 * @type {IFrom2To}
 */
export const jp2zht = _.jp2zht as IFrom2To;
/**
 * only 只將 日文漢字 => 轉為簡漢字
 * @type {IFrom2To}
 */
export const jp2zhs = _.jp2zhs as IFrom2To;

/**
 * only 簡漢字 => 轉為日漢字
 * @type {IFrom2To}
 */
export const zhs2jp = _.zhs2jp;

/**
 * only 只將 繁漢字 => 轉為日文漢字
 * @type {IFrom2To}
 */
export const zht2jp = _.zht2jp;

/**
 * only 只將簡繁日 當中共通的 繁漢字 => 轉為簡漢字
 * 請勿作為簡繁轉換用
 * @type {IFrom2To}
 */
export const zht2zhs = _.zht2zhs;

/**
 * only 只將簡繁日 當中共通的 簡漢字 => 轉為繁漢字
 * 請勿作為簡繁轉換用
 * @type {IFrom2To}
 */
export const zhs2zht = _.zhs2zht;

/**
 * only 只將簡繁日 當中共通的 簡繁漢字 => 轉為日文漢字
 *
 * @alias cjk2jp
 *
 * @param str
 * @param {IOptions} options
 * @returns {string}
 */
export function zh2jp(str, options?: IOptions): string
{
	if (!REGEXP_TEST.test(str.toString()))
	{
		return str;
	}

	options = Object.assign({}, defaultOptions, options);

	return UString.split(str, '')
		.map(function (char: string)
		{
			if (options.skip && options.skip.indexOf(char) != -1)
			{
				return char;
			}

			let c: string;
			if (c = _getdata(char, KEY_ZHT, KEY_JP, options.safe))
			{
				return c;
			}
			else if (c = _getdata(char, KEY_ZHS, KEY_JP, options.safe))
			{
				return c;
			}

			return char;
		})
		.join('')
		;
}

/**
 * only 只將簡繁日 當中共通的 簡繁漢字 => 轉為日文漢字
 *
 * @alias zh2jp
 *
 * @param str
 * @param {IOptions} options
 * @returns {string}
 */
export const cjk2jp = zh2jp;

/**
 * only 只將簡繁日 當中共通的 漢字 => 轉為繁體漢字
 * 請勿作為簡繁轉換用
 *
 * @param str
 * @param {IOptions} options
 * @returns {string}
 */
export function cjk2zht(str, options?: IOptions): string
{
	if (!REGEXP_TEST.test(str.toString()))
	{
		return str;
	}

	options = Object.assign({}, defaultOptions, options);

	return UString.split(str, '')
		.map(function (char: string)
		{
			if (options.skip && options.skip.indexOf(char) != -1)
			{
				return char;
			}

			let c: string;
			if (c = _getdata(char, KEY_JP, KEY_ZHT, options.safe))
			{
				return c;
			}
			else if (c = _getdata(char, KEY_ZHS, KEY_ZHT, options.safe))
			{
				return c;
			}

			return char;
		})
		.join('')
		;
}

/**
 * only 只將簡繁日 當中共通的 漢字 => 轉為簡體漢字
 * 請勿作為簡繁轉換用
 *
 * @param str
 * @param {IOptions} options
 * @returns {string}
 */
export function cjk2zhs(str, options?: IOptions): string
{
	if (!REGEXP_TEST.test(str.toString()))
	{
		return str;
	}

	options = Object.assign({}, defaultOptions, options);

	return UString.split(str, '')
		.map(function (char: string)
		{
			if (options.skip && options.skip.indexOf(char) != -1)
			{
				return char;
			}

			let c: string;
			if (c = _getdata(char, KEY_JP, KEY_ZHS, options.safe))
			{
				return c;
			}
			else if (c = _getdata(char, KEY_ZHT, KEY_ZHS, options.safe))
			{
				return c;
			}

			return char;
		})
		.join('')
		;
}

export default exports as typeof import('./index');
