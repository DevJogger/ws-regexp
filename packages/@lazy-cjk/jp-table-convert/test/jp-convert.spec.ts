import { jp2zht, zh2jp, cjk2zht, cjk2zhs, cjk2jp, jp2zhs } from '../lib';

describe(`describe`, () =>
{
	const input = '魔物解説　ランク等話　蚀蝕蝕王で触王 冒険者ギルド解説 蚀|蝕战|戦马|馬亚|亞國預中日漢字對照表';

	[

		zh2jp,
		jp2zhs,
		jp2zht,

		cjk2jp,
		cjk2zhs,
		cjk2zht,

	].forEach(fn => {
		test(fn.name, () =>
		{

			let actual = fn(input);

			expect(actual).not.toStrictEqual(input);

			expect(actual).toMatchSnapshot();

		});
	});

})
