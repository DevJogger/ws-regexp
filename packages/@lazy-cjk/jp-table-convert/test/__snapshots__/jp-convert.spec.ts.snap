// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe  1`] = `"魔物解说　ランク等话　蚀蚀蚀王で触王 冒险者ギルド解说 蚀|蚀战|战马|马亚|亞國預中日汉字對照表"`;

exports[`describe  2`] = `"魔物解說　ランク等話　蚀蝕蝕王で觸王 冒險者ギルド解說 蚀|蝕战|戰马|馬亚|亞國預中日漢字對照表"`;

exports[`describe cjk2zhs 1`] = `"魔物解说　ランク等话　蚀蚀蚀王で触王 冒险者ギルド解说 蚀|蚀战|战马|马亚|亚国預中日汉字对照表"`;

exports[`describe cjk2zht 1`] = `"魔物解說　ランク等話　蝕蝕蝕王で觸王 冒險者ギルド解說 蝕|蝕戰|戰馬|馬亞|亞國預中日漢字對照表"`;

exports[`describe zh2jp 1`] = `"魔物解説　ランク等話　蝕蝕蝕王で触王 冒険者ギルド解説 蝕|蝕戦|戦馬|馬亜|亜国預中日漢字対照表"`;

exports[`describe zh2jp 2`] = `"魔物解説　ランク等話　蝕蝕蝕王で触王 冒険者ギルド解説 蝕|蝕戦|戦馬|馬亜|亜国預中日漢字対照表"`;
